hack/generator/pkg/codegen/pipeline_create_arm_types.go

```diff
@@ -198,22 +199,29 @@ func transformTypeDefinition(
 	def astmodel.TypeDefinition,
 	handlers []conversionHandler) (astmodel.TypeDefinition, error) {
 
-	originalType, ok := def.Type().(*astmodel.ObjectType)
-	if !ok {
-		return astmodel.TypeDefinition{}, errors.Errorf("input type %q (%T) was not of expected type Object", def.Name(), def.Type())
+	if !astmodel.IsObjectType(def.Type()) {
+		return astmodel.TypeDefinition{}, errors.Errorf("input type %q (%T) did not contain expected type Object", def.Name(), def.Type())
 	}
 
-	resultType := originalType
-	var err error
-	for _, handler := range handlers {
-		resultType, err = handler(resultType)
-		if err != nil {
-			return astmodel.TypeDefinition{}, err
+	visitor := astmodel.MakeTypeVisitor()
+	visitor.VisitObjectType = func(this *astmodel.TypeVisitor, it *astmodel.ObjectType, ctx interface{}) (astmodel.Type, error) {
```

Comment:
I am not sure I agree that it makes sense to use a visitor here?

Effectively what you've done (I think) is:

1. Asserted we have an object or ARM type (`astmodel.IsObjectType`) above.
2. It's practically speaking always an `astmodel.ObjectType`
3. Created a visitor and overridden the visit func for object type so that it doesn't call `visit`... effectively meaning we aren't traversing anything with the visitor (I think?)

So it ends up feeling like this visitor is adding complexity but not really actually doing anything - My understanding (and I could be missing something here) is that the visitor is for managing an arbitrary `Type`. Since in this case we know what we have (and we have defined custom handlers for them which don't use the visitor framework) the visitor isn't really helping us?

Reply:
When I wrote this bit I think I played around with trying to use visitor for the handlers but couldn't get it to work cleanly... I recall having the argument in my head for why it didn't make sense, but of course I didn't write it down and now I don't remember it =P

Reply:
The content of the `TypeDefinition` may be either an `ObjectType` _or_ an `ArmType`. I tried writing the code to handle both situations in this method and it got tremendously convoluted. The `TypeVisitor` already knows how to unwrap/rewrap a definition (and an `ArmType`) so reusing that to make the transformation allowed the code to stay relatively easy to read.